htlcswitch/circuit.go

```diff
@@ -27,148 +177,547 @@ type PaymentCircuit struct {
+	if err != nil {
+		return err
+	}
+
+	// Skip encoding of error encrypter if this half add does not have one.
+	if encrypterType == EncrypterTypeNone {
+		return nil
+	}
+
+	return c.ErrorEncrypter.Encode(w)
 }
 
-// CircuitMap is a data structure that implements thread safe storage of
-// circuit routing information. The switch consults a circuit map to determine
-// where to forward HTLC update messages. Each circuit is stored with it's
-// outgoing HTLC as the primary key because, each offered HTLC has at most one
-// received HTLC, but there may be multiple offered or received HTLCs with the
-// same payment hash. Circuits are also indexed to provide fast lookups by
+// Decode reads a HalfPaymentCircuit from the provided io.Reader.
+func (c *HalfPaymentCircuit) Decode(r io.Reader) error {
+	if _, err := io.ReadFull(r, c.PaymentHash[:]); err != nil {
+		return err
+	}
+
+	var scratch [8]byte
+
+	if _, err := io.ReadFull(r, scratch[:]); err != nil {
+		return err
+	}
+	c.IncomingAmount = lnwire.MilliSatoshi(
+		binary.BigEndian.Uint64(scratch[:]))
+
+	if _, err := io.ReadFull(r, scratch[:]); err != nil {
+		return err
+	}
+	c.OutgoingAmount = lnwire.MilliSatoshi(
+		binary.BigEndian.Uint64(scratch[:]))
+
+	if _, err := io.ReadFull(r, scratch[:]); err != nil {
+		return err
+	}
+	c.IncomingChanID = lnwire.NewShortChanIDFromInt(
+		binary.BigEndian.Uint64(scratch[:]))
+
+	if _, err := io.ReadFull(r, scratch[:]); err != nil {
+		return err
+	}
+	c.IncomingHTLCID = binary.BigEndian.Uint64(scratch[:])
+
+	var encrypterType EncrypterType
+	err := binary.Read(r, binary.BigEndian, &encrypterType)
+	if err != nil {
+		return err
+	}
+
+	switch encrypterType {
+	case EncrypterTypeNone:
+		return nil
+	case EncrypterTypeSphinx:
+		c.ErrorEncrypter = NewSphinxErrorEncrypter()
+	case EncrypterTypeMock:
+		c.ErrorEncrypter = NewMockObfuscator()
+	default:
+		return UnknownEncrypterType(encrypterType)
+	}
+
+	return c.ErrorEncrypter.Decode(r)
+}
+
+// HalfKey returns the key to be used for this half-circuit.
+func (c *HalfPaymentCircuit) HalfKey() CircuitKey {
+	return CircuitKey{
+		chanID: c.IncomingChanID,
+		htlcID: c.IncomingHTLCID,
+	}
+}
+
+// CircuitMap is a data structure that implements thread safe, persistent
+// storage of circuit routing information. The switch consults a circuit map to
+// determine where to forward HTLC update messages. Each circuit is stored with
+// it's outgoing HTLC as the primary key because, each offered HTLC has at most
+// one received HTLC, but there may be multiple offered or received HTLCs with
+// the same payment hash. Circuits are also indexed to provide fast lookups by
 // payment hash.
-//
-// TODO(andrew.shvv) make it persistent
 type CircuitMap struct {
-	mtx       sync.RWMutex
-	circuits  map[circuitKey]*PaymentCircuit
-	hashIndex map[[32]byte]map[PaymentCircuit]struct{}
+	// db provides the persistent storage engine for the circuit map.
+	db *channeldb.DB
```

Comment:
Let's add a TODO here to later abstract out the storage engine? In the future, this can simply be an memory mapped on-disk hash table.